### ISSUE TITLE
Improve startup performance

### ISSUE THREAD

**Issue by Ortham** [loot/loot#1006]
Continuing from loot/loot-api#35, the slowest bits of startup are now all in the LOOT application. Example timings:

```
[18:08:15.338328] [info]: LOOT Version: 0.13.1+fc664279
[18:08:15.558051] [info]: Attempting to open link: http://loot/ui/index.html
[18:08:17.913766] [info]: Getting LOOT's supported languages.
[18:08:17.914230] [info]: Getting LOOT's settings.
[18:08:18.678810] [trace]: Sending progress update: Parsing, merging and evaluating metadata...
[18:08:20.950628] [info]: Loading 374 plugins using 6 threads, with up to 63 plugins per thread.
[18:08:21.112473] [trace]: Starting plugin loading.
[18:08:21.207521] [debug]: Parsing metadata list(s).
[18:08:21.666102] [trace]: Getting masterlist metadata for: Skyrim.esm
[18:08:22.294237] [trace]: Getting masterlist metadata for: ccbgssse010-petdwarvenarmoredmudcrab.esl
[18:08:25.455277] [info]: Getting LOOT's version.
```

Observations:

- Initialising the UI takes a couple of seconds, the Polymer 3 update might help with this.
- Loading settings is weirdly slow.
- Finding valid plugins to load is a bit slow (~ 2s), multithreading that like how plugin loading is might help.
- Getting masterlist metadata for non-DLC plugins takes 3 seconds, mostly due to CRC calculations. There's nothing particularly inefficient going on, getting metadata in parallel might help.

**Comment by Ortham** [comment]
Updated timings from the current `update-api` branch:

```
[13:50:34.540712] [info]: LOOT Version: 0.13.4+02f3853d
[13:50:34.543665] [info]: Initialising filesystem-related data for game: TES V: Skyrim Special Edition
[13:50:34.575691] [debug]: Initialising language settings in UI thread.
[13:50:34.591665] [info]: Attempting to open link: http://loot/ui/index.html
[13:50:35.254868] [info]: Getting LOOT's supported languages.
[13:50:35.255665] [info]: Getting LOOT's settings.
[13:50:35.608667] [trace]: Sending progress update: Parsing, merging and evaluating metadata...
[13:50:35.709005] [info]: Loading 401 plugins using 8 threads, with up to 51 plugins per thread.
[13:50:35.742695] [trace]: Starting plugin loading.
[13:50:36.028733] [debug]: Parsing metadata list(s).
[13:50:36.128665] [trace]: Checking that the current install is valid according to Skyrim.esm's data.
[13:50:39.138667] [info]: Getting LOOT's version.
```

Observations:

- Initialising the UI takes 0.7 s, down from 2.4 s.
- Loading settings in't actually that slow, the time between that log statement and the next also covers initialising UI text, filling in the settings table and populating dropdowns. That's not a huge amount to do, but the DOM is slow.
- Finding plugins is much faster now at 0.1 s, down from 2.3 s. This is probably due to an esplugin update that fails parsing if a group name field has an unexpected value, so parsing fails much faster for non-plugin files.
- Getting masterlist metadata hasn't changed, which is expected.

Startup time has gone from 10 seconds to 5, but it might be able to knock another couple of seconds off that by multithreading getting masterlist metadata.

**Comment by Ortham** [comment]
There was a bug introduced at some point that caused startup to load game data twice, this has been fixed in e100d7cc3be5fb42f9b15853ec0dad5de3263c92. dcfb934cb1d68a3d582819dcfb13cf84ef720761 includes much faster CRC calculation for larger files during condition evaluation, which improves startup times: I saw a ~ 7.5s startup time go down to 6s, CptMcSplody saw a ~ 2x improvement.

AL on Discord reports startup takes ~ 51s with a Data folder that has 5 plugins and ~ 40,000 files in it, here's the debug log they shared: [LOOTDebugLog.txt](https://github.com/loot/loot/files/4777179/LOOTDebugLog.txt). Breaking down the timings in it:

```
[21:09:17.138674] [info]: LOOT Version: 0.15.1+dcfb934c
...
[21:09:17.787815] [info]: Loading the current load order state.
[21:09:29.131304] [trace]: Getting active plugins.
[21:09:29.131445] [trace]: Scanning for plugins in G:\Programs\Games\SteamLibrary\steamapps\common\Skyrim Special Edition\Data
... ~40,000 lines later, mostly ".. is not a valid plugin."
[21:09:40.752337] [info]: Loading 5 plugins using 5 threads, with up to 1 plugins per thread.
[21:10:07.274096] [trace]: Starting plugin loading.
[21:10:07.280929] [debug]: Preparing to parse masterlist.
[21:10:07.281036] [debug]: Parsing metadata list(s).
[21:10:07.281128] [debug]: Loading file: C:\Users\Al12r\AppData\Local\LOOT\Skyrim Special Edition\masterlist.yaml
... lines evaluating conditions and checking if plugins are active
[21:10:07.727484] [info]: Getting whether or not LOOT was run to auto-sort.
```

So loading load order state took 11.5s and finding plugins took 11.5s. The 26.5s pause between calculating how plugin loading will be split between threads and actually loading them looks odd, but there's a call to `Game::CacheArchives()` at that point, which does another directory scan with a path equivalence check, checking every file (the other scans compare file extensions as strings).

The `CacheArchives()` check can be trivially optimised since all archive file extensions are ACSII-only, but the others may require more work. However, 40,000 files in the `Data` folder is far, far beyond any realistic scenario (with 402 Skyrim SE plugins installed, I've only got 451 items (files and folders) in my `Data` folder), so anything that makes the implementation significantly more complex almost certainly isn't worth it.

**Comment by Ortham** [comment]
Here's a collection of logs excerpts from various versions of LOOT running on the same game install. The masterlist versions vary due to varying support for masterlist syntaxes over the versions, but in each case LOOT's loading the most recent masterlist that it supports.

LOOT v0.13.1 was the latest release at the time this issue was filed:

```
[18:58:15.398400] [info]: LOOT Version: 0.13.1+fc664279
[18:58:15.401247] [info]: LOOT API Version: 0.13.5+cf9f9556
...
[18:58:16.810104] [debug]: Loading the current load order state.
[18:58:16.869849] [trace]: Scanning for plugins in D:\Games\Steam\steamapps\common\Skyrim Special Edition\Data
...
[18:58:17.011658] [info]: Loading 410 plugins using 6 threads, with up to 69 plugins per thread.
...
[18:58:17.018953] [trace]: Starting plugin loading.
...
[18:58:17.368298] [debug]: Preparing to parse masterlist.
[18:58:17.368397] [debug]: Preparing to parse userlist.
[18:58:17.368419] [debug]: Parsing metadata list(s).
...
[18:58:19.719947] [debug]: Getting load order.
```

LOOT v0.15.1 was the latest release at the time of my last comment above:

```
[19:04:33.624576] [info]: LOOT Version: 0.15.1+82c756e1
[19:04:33.625710] [info]: LOOT API Version: 0.15.1+f725dd77
...
[19:04:34.651408] [info]: Loading the current load order state.
[19:04:34.730380] [trace]: Getting active plugins.
[19:04:34.730514] [trace]: Scanning for plugins in D:\Games\Steam\steamapps\common\Skyrim Special Edition\Data
...
[19:04:34.840078] [info]: Loading 410 plugins using 6 threads, with up to 69 plugins per thread.
[19:04:34.920000] [trace]: Starting plugin loading.
[19:04:35.781876] [debug]: Preparing to parse masterlist.
[19:04:35.781973] [debug]: Preparing to parse userlist.
[19:04:35.781994] [debug]: Parsing metadata list(s).
...
[19:04:41.046153] [info]: Loading the current load order state.
[19:04:41.115838] [trace]: Getting active plugins.
[19:04:41.115914] [trace]: Scanning for plugins in D:\Games\Steam\steamapps\common\Skyrim Special Edition\Data
...
[19:04:41.229020] [info]: Loading 410 plugins using 6 threads, with up to 69 plugins per thread.
[19:04:41.311243] [trace]: Starting plugin loading.
[19:04:42.225987] [debug]: Preparing to parse masterlist.
[19:04:42.226079] [debug]: Preparing to parse userlist.
[19:04:42.226099] [debug]: Parsing metadata list(s).
[19:04:42.226169] [debug]: Loading file: C:\Users\Oliver\AppData\Local\LOOT\Skyrim Special Edition\masterlist.yaml
...
[19:04:47.529999] [info]: Getting whether or not LOOT was run to auto-sort.
```

Note that the last logged line has changed, but it represent the same point as in the log excerpt above.

LOOT v0.16.0 included the fix for the double game load and the `CacheArchives()` optimisation mentioned in my previous commit above:

```
[19:08:58.044251] [info]: LOOT Version: 0.16.0+b1094e67
[19:08:58.045389] [info]: LOOT API Version: 0.16.1+d8c9b983
...
[19:08:59.007842] [info]: Loading the current load order state.
[19:08:59.091257] [trace]: Getting active plugins.
[19:08:59.092166] [trace]: Scanning for plugins in D:\Games\Steam\steamapps\common\Skyrim Special Edition\Data
...
[19:08:59.243186] [info]: Loading 410 plugins using 6 threads, with up to 69 plugins per thread.
[19:08:59.244134] [trace]: Starting plugin loading.
[19:08:59.295254] [debug]: Preparing to parse masterlist.
[19:08:59.295526] [debug]: Preparing to parse userlist.
[19:08:59.295687] [debug]: Parsing metadata list(s).
...
[19:09:01.920401] [info]: Getting whether or not LOOT was run to auto-sort.
```

LOOT v0.17.0 is the current latest release:

```
[18:30:23.223967] [info]: LOOT Version: 0.17.0+36fedfea
[18:30:23.225233] [info]: LOOT API Version: 0.17.1+e6706057
...
[18:30:24.021306] [info]: Loading the current load order state.
[18:30:24.100636] [trace]: Getting active plugins.
...
[18:30:24.265522] [info]: Loading 410 plugins using 6 threads, with up to 69 plugins per thread.
[18:30:24.266595] [trace]: Starting plugin loading.
[18:30:24.313310] [debug]: Preparing to parse masterlist prelude.
[18:30:24.313563] [debug]: Preparing to parse masterlist.
[18:30:24.313850] [debug]: Preparing to parse userlist.
[18:30:24.314005] [debug]: Parsing metadata list(s).
...
[18:30:26.997643] [info]: Getting whether or not LOOT was run to auto-sort.
```

libloot v0.17.3 introduced caching of plugin regex objects, which significantly improves startup performance.

LOOT v0.17.0 (plus a few commits changing documentation) and libloot v0.17.3:

```
[18:34:56.556931] [info]: LOOT Version: 0.17.0+ee702c5a
[18:34:56.559923] [info]: LOOT API Version: 0.17.3+27d8c9bc
...
[18:34:57.561326] [info]: Loading the current load order state.
[18:34:57.643018] [trace]: Getting active plugins.
...
[18:34:57.792101] [info]: Loading 410 plugins using 6 threads, with up to 69 plugins per thread.
[18:34:57.792926] [trace]: Starting plugin loading.
[18:34:57.839702] [debug]: Preparing to parse masterlist prelude.
[18:34:57.840023] [debug]: Preparing to parse masterlist.
[18:34:57.840257] [debug]: Preparing to parse userlist.
[18:34:57.840412] [debug]: Parsing metadata list(s).
...
[18:34:59.656464] [info]: Getting whether or not LOOT was run to auto-sort.
```

LOOT's Qt-based UI (current `qt` branch head, not noting the commit because it'll get squashed anyway):

```
[18:46:05.627092] [info]: LOOT version: 0.17.0+519257ce
[18:46:05.629975] [info]: libloot version: 0.17.3+27d8c9bc
...
[18:46:05.910514] [info]: Loading the current load order state.
[18:46:05.994878] [trace]: Getting active plugins.
[18:46:05.995045] [trace]: Scanning for plugins in D:\Games\Steam\steamapps\common\Skyrim Special Edition\Data
...
[18:46:06.103566] [info]: Loading 410 plugins using 6 threads, with up to 69 plugins per thread.
[18:46:06.104201] [trace]: Starting plugin loading.
[18:46:06.147314] [debug]: Preparing to parse masterlist prelude.
[18:46:06.147423] [debug]: Preparing to parse masterlist.
[18:46:06.147535] [debug]: Preparing to parse userlist.
[18:46:06.147552] [debug]: Parsing metadata list(s).
...
[18:46:07.072401] [trace]: Checking if plugin "Hunterborn - Soups and Stews.esp" is active.
```

Note that the last logged line has changed, but it represent the same point as in the log excerpts above.

Summary table:

| LOOT version | Startup time |
|---------------|--------------|
|0.13.1|4.322
|0.15.1 |13.905
|0.16.0 |3.876
|0.17.0 |3.774
|0.17.0 + libloot v0.17.3 |3.010
|`qt` branch|1.445

After three and a half years, startup is about to get three times faster, so I think this can be closed once the `qt` branch is merged.

**Comment by sibir-ine** [comment]
Out of curiosity, I did a comparison of startup times between v0.17.0, the current Qt 6 64-bit build, & the current Qt 5 32-bit build & figured I'd share my results; I saw an even bigger improvement. Also of note is that the Qt 5 build isn't much slower than Qt 6, so that's good news for Windows 7 & 32-bit users.

v0.17.0:
```
[23:12:38.270187] [info]: LOOT Version: 0.17.0+36fedfea
[23:12:38.270286] [info]: LOOT API Version: 0.17.1+e6706057
...
[23:12:39.314097] [info]: Loading the current load order state.
[23:12:39.808846] [trace]: Getting active plugins.
[23:12:39.809968] [trace]: Scanning for plugins in C:\Games\Steam\steamapps\common\Skyrim Special Edition\Data
...
[23:12:40.122162] [info]: Loading 540 plugins using 12 threads, with up to 45 plugins per thread.
[23:12:40.206149] [trace]: Starting plugin loading.
[23:12:40.339781] [debug]: Preparing to parse masterlist prelude.
[23:12:40.340220] [debug]: Preparing to parse masterlist.
[23:12:40.340357] [debug]: Preparing to parse userlist.
[23:12:40.340419] [debug]: Parsing metadata list(s).
...
[23:12:55.636530] [info]: Getting whether or not LOOT was run to auto-sort.
```

Qt 5:
```
[23:18:39.734314] [info]: LOOT version: 0.17.0+af3e2a42
[23:18:39.734393] [info]: libloot version: 0.17.3+27d8c9bc
...
[23:18:40.146089] [info]: Loading the current load order state.
[23:18:41.217829] [trace]: Getting active plugins.
[23:18:41.218326] [trace]: Scanning for plugins in C:\Games\Steam\steamapps\common\Skyrim Special Edition\Data
...
[23:18:41.490866] [info]: Loading 540 plugins using 12 threads, with up to 45 plugins per thread.
[23:18:41.590939] [trace]: Starting plugin loading.
[23:18:41.734957] [debug]: Preparing to parse masterlist prelude.
[23:18:41.735029] [debug]: Preparing to parse masterlist.
[23:18:41.735085] [debug]: Preparing to parse userlist.
[23:18:41.735098] [debug]: Parsing metadata list(s).
...
[23:18:44.646519] [trace]: Checking if plugin "Hunterborn - Soups and Stews.esp" is active.
```

Qt 6:
```
[23:21:02.582360] [info]: LOOT version: 0.17.0+af3e2a42
[23:21:02.582435] [info]: libloot version: 0.17.3+27d8c9bc
...
[23:21:02.723565] [info]: Loading the current load order state.
[23:21:03.018660] [trace]: Getting active plugins.
[23:21:03.019068] [trace]: Scanning for plugins in C:\Games\Steam\steamapps\common\Skyrim Special Edition\Data
...
[23:21:03.259260] [info]: Loading 540 plugins using 12 threads, with up to 45 plugins per thread.
[23:21:03.348584] [trace]: Starting plugin loading.
[23:21:03.444060] [debug]: Preparing to parse masterlist prelude.
[23:21:03.444118] [debug]: Preparing to parse masterlist.
[23:21:03.444162] [debug]: Preparing to parse userlist.
[23:21:03.444174] [debug]: Parsing metadata list(s).
...
[23:21:05.956997] [trace]: Checking if plugin "Hunterborn - Soups and Stews.esp" is active.
```

Summary:
| LOOT version | Startup time |
| ----- | ----- |
| v0.17.0 | 17.366343
| Qt 5 | 4.912205
| Qt 6 | 3.374637

That's a little over 5 times faster for me! Also, both of our logs ended in checking `Hunterborn - Soups and Stews.esp`'s active status, which I thought was kind of funny. 😛

**Comment by Ortham** [comment]
The new UI has landed as of 6daa98c0e45eddb4d5b37312302592ab4cf476c8 with those speed improvements, so closing this.
